Allow HomeKit name to be customized

schmittx opened the pull request:
## Description:
HomeKit accessory information can be customized by specifying `name` within `entity_config`.

**Related issue (if applicable):** N/A

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5314

## Example entry for `configuration.yaml` (if applicable):
```yaml
homekit:
  port: 51825
  filter:
    include_domains:
      - lock
  entity_config:
    lock.front_door:
      name: Front Door Lock
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

schmittx commented:
@cdce8p This is my first draft, take a look and let me know your feedback. As discussed in #14114, this doesn't provide much direct functionality but just gives a cleaner experience for those that care. It's a similar concept to that of the [Alexa cloud component](https://www.home-assistant.io/cloud/alexa/).

cdce8p commented:
I have to think about it, but at the moment I'm not convinced this should be added.

cdce8p commented:
There have been a few changes with #14278
If you rebase it, I can take a new look at it next week

schmittx committed:
Initial commit

schmittx commented:
@cdce8p Rebased with the latest updates

cdce8p commented on homeassistant/components/homekit/accessories.py:
```diff
@@ -64,13 +65,18 @@ def wrapper(*args):
         """Initialize a Accessory object."""
-        super().__init__(name, aid=aid)
-        domain = split_entity_id(entity_id)[0].replace("_", " ").title()
+        display_name = config.get(CONF_NAME, name)
```
Move that call to `get_accessory`

cdce8p reviewed (changes requested):
Although I personally don't see much value in it, we can move forward with it.
Can you add tests and update the doc?

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -23,6 +24,21 @@ def validate_entity_config(values):
+        model = config.get(CONF_MODEL)
+        params[CONF_MODEL] = cv.string(model) if model else None
+
+        serial_number = config.get(CONF_SERIAL_NUMBER)
```
The empty lines between these four config opions (`name` - `serial_number`) can be removed

cdce8p commented:
Regarding the config dictionary: I was thinking about how we can improve it since I don't 
Iike adding it to every `super` call.

Why don't we use the same custruct as for `hass` and the `entity_id`? Remove the keyword in `get_accessory` and pass it through `args`. Than assign it to `self.config` in `HomeAccessory`.

schmittx committed:
Update per comments

schmittx committed:
Add util test

schmittx commented:
@cdce8p I updated per your comments, updated the doc, and updated the `test_util`. I'm not sure what other tests need to be added or updated, can you give a little advice?

cdce8p commented:
@schmittx As a starting point take a look at the failing tests: https://travis-ci.org/home-assistant/home-assistant/jobs/375445342#L808. Since `config` is no longer a keyword argument, most of the accessory tests are failing. As an example:
```py
acc = GarageDoorOpener(self.hass, 'Cover', garage_door, 2, config=None)

# Would need to change to this:
acc = GarageDoorOpener(self.hass, 'Cover', garage_door, 2)
```
Basically config can be removed, except for `security_systems`.

---

Another thing would be to add checks for the newly added vars to `test_home_accessory`: https://github.com/home-assistant/home-assistant/blob/dev/tests/components/homekit/test_accessories.py#L91-L97

---

And for the `name` config var a test in `get_accessories`, here: https://github.com/home-assistant/home-assistant/blob/dev/tests/components/homekit/test_get_accessories.py#L33

schmittx committed:
Update accessory tests

schmittx commented:
@cdce8p Thanks for the help, I updated/added most of the tests without problem. But, I keep getting KeyError with `test_get_accessory_customize_name` and `test_alarm_control_panel` as they're currently written. Can you take a look and give me some suggestions?

cdce8p reviewed (commented):
I added a few comments that should resolve the failing tests.

Although my initial doubts, this is coming together quite nicely.
Just a few notes FYI:
 - Most likely a PR will be merged soon that changes one of the `homekit` tests, which will require a small rebase #14314 
 - Even after all changes I would like to wait a few more days (the next weekend, maybe earlier) before I merge this. That way I'm still able to push bug fixes if I need to.

schmittx committed:
Updated tests

schmittx commented on tests/components/homekit/test_get_accessories.py:
```diff
@@ -21,17 +21,24 @@
         is None
 
 
+def test_get_accessory_customize_name():
```
I agree, I cleaned up both names.

schmittx commented on tests/components/homekit/test_get_accessories.py:
```diff
@@ -21,17 +21,24 @@
+    """Test with customized name."""
+    config = {CONF_NAME: 'Customize Name'}
+    acc = get_accessory(None, State('light.demo', 'on'), 2, config)
+    assert acc.display_name == 'Customize Name'
```
Awesome, thanks for the assist!

schmittx commented on homeassistant/components/homekit/util.py:
```diff
@@ -23,6 +24,18 @@ def validate_entity_config(values):
+        params[CONF_MODEL] = cv.string(model) if model else None
+        serial_number = config.get(CONF_SERIAL_NUMBER)
+        params[CONF_SERIAL_NUMBER] = \
+            cv.string(serial_number) if serial_number else None
```
Agreed, much cleaner.

schmittx commented on tests/components/homekit/test_accessories.py:
```diff
@@ -88,14 +89,29 @@ def test_home_accessory(self):
+        config = {CONF_MANUFACTURER: 'Customize Manufacturer',
+                  CONF_MODEL: 'Customize Model',
+                  CONF_SERIAL_NUMBER: 'Customize Serial Number'}
+        acc = HomeAccessory('hass', 'test_name', 'test_model.demo', 2, config)
```
Agreed.

schmittx commented:
@cdce8p Sounds good to me, this should be good to go now once the tests pass.

FYI, I'm also working on adding support for Fan entities. I should have that PR ready later this week or next weekend.

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -24,17 +24,11 @@ def validate_entity_config(values):
+                      CONF_MODEL, CONF_SERIAL_NUMBER):
+            value = config.get(index)
+            if value:
+                params[index] = value
```
I would prefer to use `key` instead.. `index` is usually used for an int counter through lists, so it might lead to confusion here.

schmittx replied:
`key` is already used in the top level at line 20, any suggestions on what else to use?

cdce8p replied:
Missed that. What do you think of changing the top level one to `entity_id` and using key here?
(At least if I haven't missed something again.)

cdce8p commented on tests/components/homekit/test_accessories.py:
```diff
@@ -89,14 +89,6 @@ def test_home_accessory(self):
-        self.assertEqual(len(acc.services), 1)
-        serv = acc.services[0]  # SERV_ACCESSORY_INFO
-        self.assertEqual(
-            serv.get_characteristic(CHAR_MODEL).value, 'Test Model')
```
I just realized that we wouldn't/can't cover this case otherwise with the accessory below.
We might have to add it back. Sorry for the confusion.

schmittx replied:
Added back and cleaned up `test_home_accessory`, hopefully you agree with my changes.

cdce8p commented:
@schmittx Regarding fan support. I think someone else was already working on it, although I haven't seen an open PR yet. Take a look at #13805.

We should't talk to much about it in this PR though, to keep it clean. Find me on `discord`.

schmittx commented:
@cdce8p Valid point, I’ll keep fan support discussion elsewhere.

schmittx committed:
Cleaned up tests and minor change to util

cdce8p commented on tests/components/homekit/test_accessories.py:
```diff
@@ -89,13 +90,17 @@ def test_home_accessory(self):
         hass.states.set('homekit.accessory', 'off')
         hass.block_till_done()
 
+        acc = HomeAccessory(hass, 'Home Accessory', 'test_model.accessory', 2,
```
Just a minor thing. If we use `test_model.acc` instead, it should fit on one line.

schmittx replied:
Done.

cdce8p commented on tests/components/homekit/test_accessories.py:
```diff
@@ -70,6 +70,7 @@ def test_home_accessory(self):
         acc = HomeAccessory(hass, 'Home Accessory', 'homekit.accessory', 2, {})
```
Subsequently this could be `homekit.acc`

schmittx replied:
Done.

cdce8p commented on tests/components/homekit/test_accessories.py:
```diff
@@ -89,13 +90,17 @@ def test_home_accessory(self):
-        self.assertEqual(acc.display_name, 'test_name')
-        self.assertEqual(acc.aid, 2)
-        self.assertEqual(len(acc.services), 1)
+        acc = HomeAccessory(hass, 'Home Accessory', 'homekit.accessory', 2,
```
This `homekit.acc` as well.

schmittx replied:
Not applicable anymore.

cdce8p commented:
I read #14330 today. It seems that for future compatibility it might be necessary to revert some parts of #14159 in order for it to work.
 - The serial number must be unique. Maybe the `aid`?
 - To still have the `entity_id` available, the change to model should probably be reverted.
 - In the end, it might just be enough to be able to customize the accessory name.

@schmittx What do you think about that? Am I missing something?

schmittx commented:
@cdce8p I read through #14330 as well as the repos that it mentioned. I understand the benefit of being able to see the history for accessories and persistent storage but I'd be hesitant to implement it due to the convoluted method required and the need to use the Eve app exclusively (if history is desired). I also think there's a high risk of regular maintenance being needed in the event that Elgato changes their services/characteristics.

I think we should focus on using the native Apple HomeKit services/characteristics in order to guarantee a consistent experience for all users, regardless of what app is used (Home, Eve, etc.).

For now, I'd vote to focus on improving the current HomeKit component (i.e. add support for more entities like `media_player` or show switches as valve/sprinklers, adding optional characteristics to existing entities, battery tracking, etc.).

If we run out of things to improve for this component and want to revisit Fakegato in the future then let's worry about the accessory information at that time (Maybe we'd make Fakegato a config option, which would then overwrite any accessory information with unique serial numbers, etc.?)

cdce8p commented:
I agree with almost all of your points. The native services and chars are just easier to maintain at the moment and the focus should be an improving support for other components (eg. Media Player, Fan, ...).

What I'm worried about however is that it's quite difficult to take a feature away once it's implemented. If we really decide to implement `fakegate`, it shouldn't come at the cost of overriding the `serial number` if it's activated (maybe through a config option).

Therefore I would prefer if we limit the configurable parameters to the `name` only and assign the others as explained before. We can always add features later, but removing them is really difficult.

schmittx commented:
I agree that it's difficult to remove a feature once it's implemented, I just struggle with limiting functionality now in the hopes that we might add something else in future.

I'll defer to you in the end though. If we limit the confutable parameters to just name, then I'd propose to close this PR and create a new one. Then we can always come back to this in the future if necessary.

cdce8p commented:
> I agree that it's difficult to remove a feature once it's implemented, I just struggle with limiting functionality now in the hopes that we might add something else in future.

Take in mind that this doesn't add any new functionality, just the option to customize some additional parts.

> If we limit the confutable parameters to just name, then I'd propose to close this PR and create a new one. Then we can always come back to this in the future if necessary.

We shouldn't do that. This PR has some great foundations in place to implement this feature. A new PR would just copy most of it. A better way might be to create a new PR once this is merge that adds the additional parameters as well and close that for the time being.

schmittx committed:
Remove manufacturer, model, and serial number customize

cdce8p commented on homeassistant/components/homekit/util.py:
```diff
@@ -23,6 +23,10 @@ def validate_entity_config(values):
             raise vol.Invalid('The configuration for "{}" must be '
                               ' an dictionary.'.format(entity))
 
+        name = config.get(CONF_NAME)
```
Can you change `key` to `entity_id` in line `19` and `20`, the `for loop`? That way we won't have this problem in the future.

schmittx replied:
No problem, done.

schmittx committed:
Fix tests and small change per comment

schmittx commented:
@cdce8p OK, everything should be good now. Rebased with #14377 and a small cleanup afterwards.

cdce8p commented:
@schmittx Will take a closer look tomorrow. Thanks for taking the time!

cdce8p commented:
@schmittx BTW: I have started looking into the complete async conversion for the `homekit` component. [homekit-async](https://github.com/home-assistant/home-assistant/compare/dev...cdce8p:homekit-async?expand=1) if you want to take a look.

Are you in the HA discord channel? Discussing it there or in direct msg might be the easiest.

cdce8p committed:
Small changes
